Add ipa_pwpolicy module

adralioh opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Add a module for modifying FreeIPA password policies

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- New Module Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
ipa_pwpolicy

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
The module is based heavily on the existing IPA modules, and extends the base IPA arguments/documentation

A couple considerations:
- The arguments all use the raw LDAP name, which are pretty long, eg `krbpwdfailurecountinterval`. I added shorter aliases based on the `ipa` command that are easier to use.

  I did it this way because this is how the existing IPA modules do it, but I can change the default name to the shorter versions if preferred.
- All of the arguments are strings instead of integers because they're passed directly to the IPA web API, which stores them as strings. I can switch them to integers if preferred.

felixfontein commented on plugins/modules/identity/ipa/ipa_pwpolicy.py:
```diff
@@ -0,0 +1,247 @@
+pwpolicy:
+    description: Password policy as returned by IPA API
+    returned: always
+    type: dict
```
A `sample:` value would be nice.

adralioh replied:
Added in 92f88b266373cf6d9e4e71625b026776485e0d4e

felixfontein reviewed (commented):
Please note that all new modules and plugins are required to have tests (unit and/or integration tests).

felixfontein commented:
Thanks for your PR! The module looks pretty polished. I'm not familiar with FreeIPA or the ipa_* modules, so there's not much I can say except that :)

adralioh commented:
I added unit tests

ready_for_review

felixfontein commented:
Cool, looks great! :)

@Akasurde @Nosmoht @fxfitz as maintainers of related modules, can you please take a look?

Andersson007 commented on plugins/modules/identity/ipa/ipa_pwpolicy.py:
```diff
@@ -0,0 +1,260 @@
+        description:
+        - Name of the group that the policy will apply to.
+        - If omitted, the global policy will be used.
+        aliases: ["group", "name"]
```
```suggestion
        aliases: ["group", "name"]
```
i would avoid aliases which is ok to add sometimes (but not initially), e.g. when a typo or something. So i'd choose one best option for every parameter.

Andersson007 reviewed (commented):
@adralioh hi, thanks for the new module!
General things like the documentation, examples, etc. LGTM
would be good to get feedback from mentioned people

Andersson007 commented:
folks from freeipa team hinted that there's a special repo https://github.com/freeipa/ansible-freeipa/
and it looks like there's a similar module already implemented https://github.com/freeipa/ansible-freeipa/blob/master/plugins/modules/ipapwpolicy.py
@adralioh did you see that? Maybe better just to add missed functionality (if it is) to the existing one?

Andersson007 commented:
we're now discussing in IRC ansible-community the possibility of deprecating this stuff and redirecting people to the special repo

adralioh commented:
The module doesn't do anything that the one in the `ansible-freeipa` repo doesn't do as far as I can tell. I wanted to add the module here because it was the only IPA functionality I needed in my playbooks that's currently missing from `community.general`.

Was there a consensus on whether or not to deprecate the IPA modules? If they're not going to be deprecated, I think it would still be worth adding this module.

felixfontein commented on plugins/modules/identity/ipa/ipa_pwpolicy.py:
```diff
@@ -0,0 +1,260 @@
+short_description: Manage FreeIPA password policies
+description:
+- Add, modify, and delete password policy within IPA server.
+version_added: 1.3.0
```
```suggestion
version_added: 2.0.0
```

adralioh replied:
Changed in cbfd9b1

felixfontein commented:
Since there hasn't really been any discussions, it might be better to add the module here for now. @Andersson007 what do you think?

Andersson007 commented:
> Since there hasn't really been any discussions, it might be better to add the module here for now. @Andersson007 what do you think?

Ok

Andersson007 commented on plugins/modules/identity/ipa/ipa_pwpolicy.py:
```diff
@@ -0,0 +1,260 @@
+        aliases: ["lockouttime"]
+        type: str
+extends_documentation_fragment:
+- community.general.ipa.documentation
```
```suggestion
- community.general.ipa.documentation
notes:
- Supports C(check_mode).
```

adralioh replied:
Added in 09bca67

Andersson007 commented on plugins/modules/identity/ipa/ipa_pwpolicy.py:
```diff
@@ -0,0 +1,260 @@
+
+RETURN = r'''
+pwpolicy:
+    description: Password policy as returned by IPA API
```
```suggestion
    description: Password policy as returned by IPA API.
```

adralioh replied:
Added in ec7ee77

Andersson007 commented:
@adralioh thanks for the work! Let's continue here then.
my comment about aliases is still relevant.

adralioh commented:
@Andersson007 Regarding the aliases, I'm not sure which ones are the best option to keep.

The long names that I have as the default option name, eg `krbpwdfailurecountinterval`, are the raw LDAP name as returned by the IPA API. \
The shorter aliases, eg `failinterval`, are the names used on the commandline by the `ipa` command.

The other IPA modules mostly only use the LDAP name, but they're not as long as the ones in this module, so are more manageable.

Also, for reference, the module in the `ansible-freeipa` repo has aliases for both names.

I'm leaning towards removing the shorter aliases so it's consistent with the other modules, but I wanted to see what you thought.

Andersson007 commented on plugins/modules/identity/ipa/ipa_pwpolicy.py:
```diff
@@ -0,0 +1,262 @@
+options:
+    cn:
+        description:
+        - Name of the group that the policy will apply to.
+        - If omitted, the global policy will be used.
```
```suggestion
        - Name of the group that the policy will apply to.
        - If omitted, the global policy will be used.
```
Also try to use the present tense where applicable (it's from the development documentation as well)

adralioh replied:
Changed in 8349dcd20a4c7b248eb3952a5dfed93c125e31b8

Andersson007 commented:
@adralioh IMO we could use the short forms because as i can see they don't intersect with each other but we could put the long forms to descriptions, e.g. (very roughly):
```
    maxfile:
        description: Maximum password lifetime (in days). Corresponds to the C(krbmaxpwdlife) raw LDAP name.
```
BTW there's a typo in `maxfile`. I think `maxpwdlife` would sound better.

Andersson007 commented:
> @adralioh IMO we could use the short forms because as i can see they don't intersect with each other but we could put the long forms to descriptions, e.g. (very roughly):
> 
> ```
>     maxlife:
>         description: Maximum password lifetime (in days). Corresponds to the C(krbmaxpwdlife) raw LDAP name.
> ```
> 
> I think `maxpwdlife` would sound better.
> 

, providing that to know the raw names can be important for anyone. If not, i'd leave only the short forms

adralioh commented:
@Andersson007 Thanks, I changed them to short names, and removed mention of the LDAP long names since they probably aren't important.

I renamed some of the short names to be more descriptive, such as `maxlife` to `maxpwdlife` like you mentioned, and `maxfail` to `maxfailcount`.

I also improved the description of some of the options so that they're better worded.

Andersson007 commented:
@adralioh thanks for the new module!
@felixfontein thanks for reviewing!

Andersson007 commented:
@adralioh i'll add you as a maintainer to the special BOTMETA.yml file next week to keep you informed about related issues / PRs
